### PR TITLE
fix(convert-formulas.sh): add -_ to allowed chars in formula name

### PR DESCRIPTION
**Pull request by dafyddj**
### PR progress checklist (to be filled in by reviewers)

- [ ] Changes to documentation are appropriate (or tick if not required)
- [ ] Changes to tests are appropriate (or tick if not required)
- [ ] Reviews completed

---

#### Primary type

- [ ] `[build]`    Changes related to the build system
- [ ] `[chore]`    Changes to the build process or auxiliary tools and libraries such as documentation generation
- [ ] `[ci]`       Changes to the continuous integration configuration
- [ ] `[feat]`     A new feature
- [x] `[fix]`      A bug fix
- [ ] `[perf]`     A code change that improves performance
- [ ] `[refactor]` A code change that neither fixes a bug nor adds a feature
- [ ] `[revert]`   A change used to revert a previous commit
- [ ] `[style]`    Changes that do not affect the meaning of the code (white-space, formatting, missing semi-colons, etc.)

#### Secondary type

- [ ] `[docs]`     Documentation changes
- [ ] `[test]`     Adding missing or correcting existing tests

### Does this PR introduce a `BREAKING CHANGE`?

No.

### Describe the changes you're proposing

Allow '-' and '_' in formula names. Formulas already exist with '-' in the name (prior to `-formula`),
so it seems reasonable to allow creation of new ones.  

### Documentation checklist

- [x] Updated the `README` (e.g. `Available states`).
- [ ] Updated `pillar.example`.

### Testing checklist

- [ ] Included in Kitchen (i.e. under `state_top`).
- [x] Covered by new/existing tests (e.g. InSpec, Serverspec, etc.).
- [ ] Updated the relevant test pillar.

### CODE REVIEW

**Review by vutny — APPROVED** [review]
Looks good

**Review by baby-gnu — APPROVED** [review]

**Comment by baby-gnu** [comment]
I'm wondering if the tooling accept to have a `-` in the name of the formula. I think @myii told me something about this.

**Comment by noelmcloughlin** [comment]
I would be against allowing '_' (underscore) in formula names because it invalidates current assumptions around semantic naming. Some core maintainers (i.e. me) expect a standard semantic name of `<component>-formula` using 'dash' as the historic community convention for formula name.  In https://github.com/saltstack-formulas/salter I assume '-' and can add '_' but increased flexibility in the naming convention introduces complexity in other places.

**Review by daks — APPROVED** [review]

**Comment by dafyddj** [comment]
> I'm wondering if the tooling accept to have a `-` in the name of the formula. I think @myii told me something about this.

There are at least [13 formulas](https://github.com/saltstack-formulas/.github/blob/master/CONTRIBUTING.rst#semantic-release-formulas) that have `-` in their names, so I don't believe there is a tooling issue (although I welcome correction on that), and this change just brings the `convert-formula.sh` into line with current practice.

I add '_' for flexibility, but perhaps it is not necessary.

**Comment by myii** [comment]
We've currently got two formulas using an underscore in the name:

* https://github.com/saltstack-formulas/logstash_forwarder-formula
* https://github.com/saltstack-formulas/pam_access-formula

**Comment by myii** [comment]
> I'm wondering if the tooling accept to have a `-` in the name of the formula. I think @myii told me something about this.

This is a link to that conversation.  We can manage (and pretty much have to, due to the sheer volume of formulas already doing this).

https://freenode.logbot.info/saltstack-formulas/20200325#c3478581-c3478644

> -|-|-
> :-:|---|---
> 20:54 | nebuchadnezzar | a question, is there a reason to limit the new name used by ./bin/convert-formula.sh ?
> 20:54 | myii | What limit is that? Let me look at it again...
> 20:54 | nebuchadnezzar | I would like to create a new formula for go-wisper/go-carbon/go-carbonapi and this limitation is too restrictiv for me
> 20:55 | nebuchadnezzar | erf, it's not go-carbonapi it's just carbonapi ;-)
> 20:55 | myii | But where will it be located in GitHub (or other hosting service)?
> 20:56 | myii | `https://;project;/https://;repo;` right?
> 20:56 | myii | CC: @​dafydd -- you may be interested to come back to this little discussion.
> 20:57 | nebuchadnezzar | I hope it will be at saltstack-formula/go-wisper-formula
> 20:59 | nebuchadnezzar | There are 51 formulas with a dash in their name
> 20:59 | myii | OK, maybe the regex is a little restrictive.
> 20:59 | myii | But it does make difficulties, though.
> 21:00 | myii | Take the `syslog-ng-formula` for example. I remember there being some horrible things we had to do to get it working in the CI, including changing those dashes to underscores.
> 21:01 | nebuchadnezzar | ho, I was not aware of that
> 21:01 | myii | https://github.com/myii/ssf-formula/blob/master/ssf/formulas.yaml#L2130-L2156
> 21:01 | myii | You can see the underscores shown in some of the places there.
> 21:02 | myii | Maybe it could have been avoided with better design, I'm not sure. I didn't convert that formula.
> 21:07 | nebuchadnezzar | thanks, I'll start locally and I'll change to a better name after, thanks
> 21:08 | myii | You're welcome. If you figure out how to get it working with the dash instead of using an underscore, let me know. That will come in useful when converting some of the other formulas.

**Comment by myii** [comment]
@dafyddj I just need to get around testing both additions locally and then we get this merged if all is OK.  Actually, we really should update the test to have a formula name that contains both `-` and `_`, i.e. modify this section:

https://github.com/saltstack-formulas/template-formula/blob/e668832b76604b3890b7f514fa36105314145382/.travis.yml#L124-L135

How about changing from `converted-formula` to something like `test-the_use-this-template_button-formula`?

**Comment by dafyddj** [comment]
> I would be against allowing underscore in formula names because it invalidates current assumptions around semantic naming. Some core maintainers (i.e. me) expect a standard semantic name of `<component>-formula` using 'dash' as the historic community convention for formula name. In https://github.com/saltstack-formulas/salter I assume '-' and can add '_' but increased flexibility in the naming convention introduces complexity in other places.

Are you referring here specifically to the `-formula` part of the name because I do not expect anyone to use `_formula`.
Otherwise, I looked at `salter` (which I wasn't previously aware of) and can't immediately see where the underscore would cause a problem.

**Comment by dafyddj** [comment]
> How about changing from `converted-formula` to something like `test-the_use-this-template_button-formula`?

Good idea.

**Comment by baby-gnu** [comment]
I would like to propose some formulas we used at work and one of them would be `pam-mount-formula`.

File: bin/convert-formula.sh [hunk 1]
```diff
@@ -52,6 +52,11 @@ convert_formula() {
   git rm --quiet bin/convert-formula.sh AUTHORS.md CHANGELOG.md \
     docs/_static/css/custom.css docs/AUTHORS.rst docs/CHANGELOG.rst \
     docs/conf.py docs/CONTRIBUTING_DOCS.rst docs/index.rst
+  tag_out=$(git tag --list | xargs git tag --delete)
+  if [ "${DEBUG:-false}" = "true" ]; then
+    echo "$tag_out"
+  fi
+
```
Comment (baby-gnu) [inline bin/convert-formula.sh:59]: Thanks a lot :+1:

**Review by baby-gnu — APPROVED** [review]

File: bin/convert-formula.sh (hunk 1 above)
Comment (baby-gnu) [reply bin/convert-formula.sh:59]: but someone may argue that it's not the purpose of this PR ;-)

**Comment by myii** [comment]
@dafyddj I've taken this PR and rebased on the latest version of this formula.  I've then run it with:

```console
$ bin/convert-formula.sh test_the_use-this-template-button
```

I've grabbed the diff between the resultant first and second commits into a [gist](https://gist.github.com/myii/e3c49e834b6d2a90ccb6398722df9636) for commenting purposes (not necessarily action items, primarily discussion).  Working from top-to-bottom (not prioritised):

1. [[L49-L1330](https://gist.github.com/myii/e3c49e834b6d2a90ccb6398722df9636#file-test_the_use-this-template-button-diff-L49-L1330)] **General comment, not to be tackled by this PR**: A couple of files deleted with 1,250+ lines removed -- could/should this be done in the first commit rather than the second, to avoid clutter in the git history?  Further sections include:
    - [[L1416-L3287](https://gist.github.com/myii/e3c49e834b6d2a90ccb6398722df9636#file-test_the_use-this-template-button-diff-L1416-L3287)] -- another 1,850+ lines removed.
    - [[L3465-L3695](https://gist.github.com/myii/e3c49e834b6d2a90ccb6398722df9636#file-test_the_use-this-template-button-diff-L3465-L3695)] -- another 200+ lines removed.
    - So a total of 3,300+ lines removed, which could be avoided entirely from the git history of the converted repo?
1. [[L3295-L3297](https://gist.github.com/myii/e3c49e834b6d2a90ccb6398722df9636#file-test_the_use-this-template-button-diff-L3295-L3297)] **Wish-list, not for this PR, nor critical in any way**: How about the script being clever enough to modify the heading symbols to match the change in line length throughout the whole of `README.rst`?
1. [[L4332-L4334](https://gist.github.com/myii/e3c49e834b6d2a90ccb6398722df9636#file-test_the_use-this-template-button-diff-L4332-L4334)] **Change requested for this PR, as discussed in Slack**: Please use `# yamllint disable-line rule:line-length` instead of the current method; the diff here shows that fix already applied.
1. [[L4391-L4392](https://gist.github.com/myii/e3c49e834b6d2a90ccb6398722df9636#file-test_the_use-this-template-button-diff-L4391-L4392)] **Confirming behaviour throughout the Jinja in the file conversions (only the first example linked from the gist)**: So both of these apply?
    - Each hyphen `-` converted to a single underscore `_` -- this is necessary for Jinja variables, IIRC from our conversations.
    - Each underscore `-` converted to a double underscore `__` -- this one _initially feels_ a little wonky but probably makes sense overall.

All-in-all, a fantastic bit of work, kudos!

**Comment by dafyddj** [comment]
> 3\. [[L4332-L4334](https://gist.github.com/myii/e3c49e834b6d2a90ccb6398722df9636#file-test_the_use-this-template-button-diff-L4332-L4334)] **Change requested for this PR, as discussed in Slack**: Please use `# yamllint disable-line rule:line-length` instead of the current method; the diff here shows that fix already applied.

Done I'll address the other points elsewhere.

**Comment by dafyddj** [comment]
> [[L4391-L4392](https://gist.github.com/myii/e3c49e834b6d2a90ccb6398722df9636#file-test_the_use-this-template-button-diff-L4391-L4392)] **Confirming behaviour throughout the Jinja in the file conversions (only the first example linked from the gist)**: So both of these apply?
> * Each hyphen `-` converted to a single underscore `_` -- this is necessary for Jinja variables, IIRC from our conversations.
> * Each underscore `-` converted to a double underscore `__` -- this one _initially feels_ a little wonky but probably makes sense overall.

The only manipulation done is to convert a hyphen to a double underscore:
`NEW_NAME_PYSAFE=$(echo "$NEW_NAME" | sed 's/-/__/g')`
I chose a double underscore (rather than a single) just to retain some information that a change was made.

**Review by myii — APPROVED** [review]

**Comment by myii** [comment]
Thanks for all of the reviews, guys.

Merged, @dafyddj.

**Pull request merged**